py/runtime.c

```diff
@@ -1051,20 +1051,28 @@ void mp_load_method_maybe(mp_obj_t obj, qstr attr, mp_obj_t *dest) {
     } else if (attr == MP_QSTR___class__) {
         // a.__class__ is equivalent to type(a)
         dest[0] = MP_OBJ_FROM_PTR(type);
+        return;
 #endif
 
     } else if (attr == MP_QSTR___next__ && type->iternext != NULL) {
         dest[0] = MP_OBJ_FROM_PTR(&mp_builtin_next_obj);
         dest[1] = obj;
+        return;
 
     } else if (type->attr != NULL) {
         // this type can do its own load, so call it
         type->attr(obj, attr, dest);
+        if (dest[0] != MP_OBJ_NULL) {
+            return;
+        }
+
+    }
 
-    } else if (type->locals_dict != NULL) {
-        // generic method lookup
-        // this is a lookup in the object (ie not class or type)
-        assert(type->locals_dict->base.type == &mp_type_dict); // MicroPython restriction, for now
+    if (type->locals_dict != NULL) {
```

Comment:
For user defined types (instances) the `locals_dict` will now be searched twice if something is not found: once in the custom `mp_obj_instance_attr` and then a second time here.  I don't know if this has any side effects, but at the very least it's inefficient.